Show autosave message when autosave exists

adamsilverstein opened the pull request:
## Description
Enables heartbeat based autosave and connection monitoring

## How Has This Been Tested?
### Autosaves
* Create a local post in the classic editor; create an autosave revision, making some changes, letting autosaave fire, then exiting the screen.
* Open post in Gutenberg.

## Screenshots (jpeg or gifs if applicable):
Opening a post with an existing autosave:
![](https://cl.ly/2Z3T2R2s1a1p/Image%202018-01-07%20at%202.13.04%20PM.png)
Click 'view the autosave' takes you to the revisions screen where the autosave is selected and you can restore it.

Network disconnected notice (note save and switch to draft disabled):
![](https://cl.ly/2K3v3n030c1l/Image%202018-01-07%20at%208.41.32%20PM.png)
  
## Types of changes
* Check for autosave at load
* show autosave message if present
## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation. # #

adamsilverstein commented:
I'd like some design input here on two things:

**First, the post locking modal.** 
This pops up when you try to open a post someone is editing, or someone takes over a post you are editing. It looks something like this: https://codex.wordpress.org/Post_Locking

When a user takes over a post you are editing, the modal pops up and you can no longer edit, the only option is to return to the posts list screen.

When you try to open a post someone is editing, the modal offers 'go back', 'preview' and 'take over' - the editor is inactive unless you choose to take over.

**Second - the autosave confirmation message**
When a heartbeat autosave fires, the server returns a helpful string about the autosave. In the classic editor this is displayed in the footer of the editor:

![](https://cl.ly/1a0T1a2v1R2K/Edit_Post__WordPress_Local_Development...__WordPress_2018-01-02_20-52-29.jpg)

How do we want to show this save time/status in Gutenberg?

During the save action itself, I am following the classic editor and showing the publish button in a 'saving' state (also disabling them) - this helps prevent save collisions and clues the user in to the autosave happening.

adamsilverstein commented:
One more design item: heartbeat/autosave also tracks the connection status, and autosaves the data to localstorage when you are offline. do we want to show an alert with a disconnection message similar to where we show the autosave message?

adamsilverstein commented:
cc: @jasmussen @karmatosed for design input

mtias commented:
Thanks for working on this. Also looping @gziolo for the collaborative editing work. Ideally when you encounter a locked post you'll see the option to request to collaborate.

jasmussen commented:
Sorry for not looking sooner. I'll be back at this tomorrow and I'll have a look. Thanks for working on this!

jasmussen commented:
Thanks so much for working on this. It works great.

> When a user takes over a post you are editing, the modal pops up and you can no longer edit, the only option is to return to the posts list screen.
> When you try to open a post someone is editing, the modal offers 'go back', 'preview' and 'take over' - the editor is inactive unless you choose to take over.

Given we'd love to have collaborative editing, like what Matías mentions, it seems like we should either _not tackle the post locking as part of this PR_, or keep the UI minimal and virtually identical to what's there now. 

> When a heartbeat autosave fires, the server returns a helpful string about the autosave. In the classic editor this is displayed in the footer of the editor:
> How do we want to show this save time/status in Gutenberg?

Do we truly need this when we have this?

If yes, then I'd show it in the revisions UI:

<img width="301" alt="screen shot 2018-01-04 at 11 00 16" src="https://user-images.githubusercontent.com/1204802/34558447-7e21a736-f13e-11e7-87e3-d0c7a28468c7.png">

It could say "2 Revisions  Draft saved 8:52:02pm" perhaps?

> One more design item: heartbeat/autosave also tracks the connection status, and autosaves the data to localstorage when you are offline. do we want to show an alert with a disconnection message similar to where we show the autosave message?

Yes, solid idea. Can we show a "warning" colored prompt like the publish prompt?

<img width="1164" alt="screen shot 2018-01-04 at 11 01 52" src="https://user-images.githubusercontent.com/1204802/34558512-b7e44744-f13e-11e7-8b08-7a89b36a6b13.png">

karmatosed commented:
I will add a little note that on observing users I have seen page locking to be a huge issue. One particular situation I recall someone saying they had to Slack someone to get page's unlocked.

I would second what @jasmussen has said and not tackle post locking. I actually think collaborative editing holds a solution that could stop a lot of experience issues.

Failing that I also agree keeping what we have now would work. I think an alert to disconnection is good, although maybe a notice color not green as it's a notice?

youknowriad commented:
> I would second what @jasmussen has said and not tackle post locking. I actually think collaborative editing holds a solution that could stop a lot of experience issues.

Agreed having collaborative editing would be great but right now when I write a post with Gutenberg I can't really share it with someone else while still a draft because if he opens the same post in Gutenberg, I might loose edits I'll be doing because of auto-saving enabled in both sides. I'd personally appreciate post locking if it's simple enough and removable once we have collaborative editing.

youknowriad commented:
> I would second what @jasmussen has said and not tackle post locking. I actually think collaborative editing holds a solution that could stop a lot of experience issues.

Agreed having collaborative editing would be great but right now when I write a post with Gutenberg I can't really share it with someone else while still a draft because if he opens the same post in Gutenberg, I might loose edits I'll be doing because of auto-saving enabled in both sides. I'd personally appreciate post locking if it's simple enough and removable once we have collaborative editing.

gziolo commented:
Please also keep in mind that Collaborative Editing is going to be disabled by default and working using WebRTC only for 2 collaborators, so I would suggest exploring flows that enable post locking, too.

gziolo commented:
@adamsilverstein it would be great to integrate at some point hooks into core and retrigger jquery events using actions 😃

adamsilverstein commented:
@jasmussen Thanks for the feedback. I like your idea of putting the last autosave status near the revisions indicator, I'll give that a try. You asked if its needed give that we already have a 'Saved' indicator - the short answer is I think this is slightly different - "Saved" means you have actually saved your post and made no further changes. An autosave is a temporary store of your changes since the last save and only used if you exit the editor without saving and then later revisit the editor.

I will remove post locking from the PR for now since there doesn't seem to be consensus on how it should work and I'd like to get autosaving in. If/when we do land post locking, we still need a design for a modal that pops up to show the interruption/locking status. Personally I see post locking as a critical feature, and if collaborative editing is enabled, maybe that can be one of the modal options? eg options are 'collaborate,  'go back', 'preview' and 'take over' when you open a locked post

adamsilverstein commented:
> I will add a little note that on observing users I have seen page locking to be a huge issue. One particular situation I recall someone saying they had to Slack someone to get page's unlocked.

@karmatosed you mean the current post locking in core, right? I guess the option to take over the post is only for admins or not that users caps anyway. Do you think post locking solves more problems (overwriting others changes) than it causes?

adamsilverstein commented:
>  I might loose edits I'll be doing because of auto-saving enabled in both sides. I'd personally appreciate post locking if it's simple enough and removable once we have collaborative editing.

@youknowriad - I agree post locking is important, and yes without it two users can open the same post and overwrite each others changes by **doing a normal save**.  Clarification however that autosaves will not overwrite other user's autosaves - each user's autosaves are kept separately (one autosave per post per user). 

We can't really add post locking until we have a design/implementation for the modal that pops up when you open a locked post - the feature is incomplete without that.

adamsilverstein commented:
> it would be great to integrate at some point hooks into core and retrigger jquery events using actions

For sure! and that way our actions would have priority so plugins could for example, respond to the data for an autosave by hooking in before our save handler.

jasmussen commented:
> You asked if its needed give that we already have a 'Saved' indicator - the short answer is I think this is slightly different - "Saved" means you have actually saved your post and made no further changes. An autosave is a temporary store of your changes since the last save and only used if you exit the editor without saving and then later revisit the editor.

Does the user need to understand the difference enough to change the label there? The idea of putting the extra information near the revisions UI is exactly because it _feels_ like an advanced distinction that isn't super meaningful to most people. Not that it sounds like we're disagreeing here, just elaborating. 

> I will remove post locking from the PR for now since there doesn't seem to be consensus on how it should work and I'd like to get autosaving in. If/when we do land post locking, we still need a design for a modal that pops up to show the interruption/locking status. Personally I see post locking as a critical feature, and if collaborative editing is enabled, maybe that can be one of the modal options? eg options are 'collaborate, 'go back', 'preview' and 'take over' when you open a locked post

When I tried this PR with two different users (one in an incognito mode), I didn't see any post-locking UI. But given what @gziolo suggested here ...

> Please also keep in mind that Collaborative Editing is going to be disabled by default and working with WebRTC only for 2 collaborators, so I would suggest exploring flows that enable post locking, too.

... maybe we shouldn't remove the post locking UI after all. 

I mean, it doesn't have to be in this PR — if you like I can do a mockup first, or if you like you can implement some barebones feature and I can CSS it up for you?

adamsilverstein commented:
> Does the user need to understand the difference enough to change the label there? 

Possibly not - good point. In the current (or previous) PR I do use the same string/spot to indicate that an autosave is happening 'Autosaving'... 'Saved'... maybe that is all we need.

>When I tried this PR with two different users (one in an incognito mode), I didn't see any post-locking UI.

The only thing in place so far is the locking itself, you should see this show up as a lock icon/status indicator in the post list screen. if you try opening the post in the classic editor you should see the locking warning modal. If not, something may not be working.

> ... maybe we shouldn't remove the post locking UI after all.

> I mean, it doesn't have to be in this PR — if you like I can do a mockup first, or if you like you can implement some barebones feature and I can CSS it up for you?

Yes, a mock up would be great, are there any other examples in gutenberg of a modal display? Yes, I'd like to break off post locking to a separate PR and change this one to focus on autosave only, which is closer to being ready.

aduth commented on editor/components/provider/index.js:
```diff
@@ -52,6 +52,9 @@ class EditorProvider extends Component {
 		// Assume that we don't need to initialize in the case of an error recovery.
 		if ( ! props.recovery ) {
 			this.store.dispatch( setupEditor( props.post, this.settings ) );
+			if ( props.autosave ) {
```
Is there ever a case where `props.recovery && props.autosave` ? Do we need to nest this?

adamsilverstein replied:
Probably not, un-nested in https://github.com/WordPress/gutenberg/pull/4218/commits/1ee973e2e2d36f4e550d5fac9582ff1c71fc03ea

aduth commented on editor/index.js:
```diff
@@ -78,14 +70,16 @@ export function recreateEditorInstance( target, settings ) {
+export function createEditorInstance( id, post, settings, autosave ) {
 	const target = document.getElementById( id );
 	const reboot = recreateEditorInstance.bind( null, target, settings );
+	setupHeartbeat();
```
Would this be able to handle multiple editors on the same page? To that point, does our revised approach of a singleton editor store support this at all? (cc @youknowriad)

adamsilverstein replied:
Mainly heartbeat should only be initialized once, but yes - autosaving should work with multiple editors - its their listeners that 'trigger' the autosave by attaching and responding to data attached to the heartbeats - each editor would handle its own local data/state...

On merging to core, we may want to change how this initialized and consider adding the autosave customizations into autosave.js directly.

youknowriad replied:
Yeah, I guess multiple editor do not work, we should rename this function `initializeEditor` or something which suggested you can't call it twice for multiple instances. (not related to this PR though)

adamsilverstein commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,173 @@
+	 * @returns {string} A concatenated string with title, content and excerpt.
+	 */
+	const getCompareString = function( state ) {
+		return ( getEditedPostTitle( state ) || '' ) + '::' + ( getEditedPostContent( state ) || '' ) + '::' + ( getEditedPostExcerpt( state ) || '' );
```
sure, thats much cleaner and easier to read, I'll update it. This function was copied from autosave.js and changed to use the editor state rather than reading from the dom.

adamsilverstein replied:
Updated in https://github.com/WordPress/gutenberg/pull/4218/commits/838f78fda82c326205a34a888f5c6d90889c2e88

adamsilverstein commented:
Here is what the autosave alert looks like when you open a post with an existing autosave:

![](https://cl.ly/2Z3T2R2s1a1p/Image%202018-01-07%20at%202.13.04%20PM.png)

Click 'view the autosave' takes you to the revisions screen where the autosave is selected and you can restore it.

adamsilverstein commented:
@jasmussen noticed the 'Saved' area for published posts and now shows 'Switch to Draft'. That seems way to prominent for such a rarely used action, should that be a bit harder to reach? Also, not sure where to show the autosaving status for published posts. For now, the only indication is the publish button gets the active blue stripes (same as 'isSaving') for around 5 seconds when the autosave data is attached to a heartbeat - this also disables the update action and helps prevent save collisions.

adamsilverstein commented:
I started working on local storage autosaving is a separate branch(https://github.com/WordPress/gutenberg/compare/feature/local-saving-for-heartbeat-autosaves?expand=1) - For now, I'd like to get this PR merged so we have autsaving actually working in Gutenberg.

adamsilverstein commented:
Network disconnected notice (note save disabled):

![](https://cl.ly/2K3v3n030c1l/Image%202018-01-07%20at%208.41.32%20PM.png)

jasmussen commented:
> noticed the 'Saved' area for published posts and now shows 'Switch to Draft'. That seems way to prominent for such a rarely used action, should that be a bit harder to reach?

I believe @mtias has some background on the value of this action, and perhaps @karmatosed can add this to the list of things to test. But the argument for putting this here is that you only see it once the post has been _published_. At this point, the value of showing "Saved" prominently in the editor bar (especially if we can show recent local storage saves in the Revisions UI in the sidebar) feels reduced, as one would think edits to published posts are rarer than edits to unpublished posts, and because "Switch to Draft" seems to _imply_ that the content is saved. 

> Also, not sure where to show the autosaving status for published posts. For now, the only indication is the publish button gets the active blue stripes (same as 'isSaving') for around 5 seconds when the autosave data is attached to a heartbeat - this also disables the update action and helps prevent save collisions.

Given the hypothesis that updating published posts happens slightly more rarely than non-published posts, is it okay if we relegate this info to the revisions UI in the sidebar?

adamsilverstein commented:
> Given the hypothesis that updating published posts happens slightly more rarely than non-published posts, is it okay if we relegate this info to the revisions UI in the sidebar?

I think so, I'll work on adding it there. That said, I don't think this is a blocker for merging, we could add this later.

This should be ready for some additional review from @aduth or @youknowriad - thanks!

youknowriad commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,173 @@
+			getEditedPostTitle( state ),
+			getEditedPostContent( state ),
+			getEditedPostExcerpt( state ),
+		] ).join( '::' );
```
Can't we use a flag like `isAutoSaveDirty` or something instead of this. It seems a bit fragile as it doesn't account for all the potential changes. I think we can use the `withChangeDetection` Helper to add a `isAutoSaveDirty`. We'd have to update the `withChangeDetection` helper to make the "isDirty" flag name customizable) cc @aduth

adamsilverstein replied:
that would make sense if possible. Originally I was tracking the autosaves in Redux, however since its currently only used here in heartbeat, I wasn't sure we needed to do that.

adamsilverstein replied:
@youknowriad I worked on adding this, storing the autosave when heartbeat fires and using withChangeDetection and isDirty to determine if its changed. let me know if this looks better

youknowriad commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,173 @@
+		// Block autosaving for 10 seconds.
+		wp.autosave.server.tempBlockSave();
+
+		const postData = getCurrentPost( state );
```
I believe this returns the last saved post and if I understand correctly we want to save the last edits here which are not included in that object. What am I missing here?

adamsilverstein replied:
oh, thats a mistake then - this should be getting the current editor content; what should I be calling to get that?

youknowriad replied:
something like that 

```
		const post = getCurrentPost( state );
		const edits = getPostEdits( state );
		const toSend = {
			...edits,
			content: getEditedPostContent( state ),
			id: post.id,
		};
```

see the `REQUEST_POST_UPDATE` effect

adamsilverstein replied:
I simplified the data setup here in https://github.com/WordPress/gutenberg/pull/4218/commits/158bbd4b4a686f67003d43e7bdb794d608a730e9, we don't need the full post object for the autosave.

youknowriad commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,173 @@
+		// Add the nonce to validate the request.
+		postData._wpnonce = jQuery( '#_wpnonce' ).val() || '';
+
+		return postData;
```
This is clearly an "effect", so it's kind of weird that we have to do this here but I guess The heartbeat API is already working this way in the classic editor and this avoids some extra work if we want to this the "redux" way.

adamsilverstein replied:
Yea, this is a bit of an artifact of glueing Gutenberg onto an existing WordPress JavaScript API

adamsilverstein replied:
ps. I'm attaching the nonce after triggering the events to match the existing behavior.

adamsilverstein commented on editor/store/reducer.js:
```diff
@@ -485,6 +485,60 @@ export function blocksMode( state = {}, action ) {
+ * @param  {Object} action Dispatched action
+ * @return {Object}        Updated state
+ */
+export const autosave = withChangeDetection( function( state = false, action ) {
```
yes, that is how i am using it, i am tracking the autosave changes using this wrapper. check the updated code from today which is a little more refined. I kept autosave outside editor in the reducer because it was cleaner that way, could possible be combined.

adamsilverstein commented:
@jasmussen In my latest commits I added the autosave message returned from the server into the state.

I added a component named AutosaveMessage that displays the autosave message when available. I tried adding in the revisions area, but remembered that only shows when there is a stored revision. for new posts you can have autosaving but no stored revisions, so that placement wasn’t great.

I then tried moving the status indicator right below the publish date. This looks better, but could use some design attention Here is how it looks now:

![](https://cl.ly/1w1W0r1g2T3h/Image%202018-01-18%20at%209.02.12%20AM.png)

 maybe it could just be smaller and lighter? The word 'Autoupdate' can probably change? 

One other thing I noticed here: the time displayed as 'published' is off by a couple of hours, not sure where thats coming from but it is currently 9am and you can see from the screen shot the autosave shows the correct time

adamsilverstein commented on editor/store/reducer.js:
```diff
@@ -485,6 +485,64 @@ export function blocksMode( state = {}, action ) {
+	}
+
+	return state;
+}, { resetTypes: [ 'UPDATE_AUTOSAVE_STATUS_MESSAGE' ] } );
```
@youknowriad when the autosave completes and the status is updated, the state.autosave.isDirty is reset to false

adamsilverstein commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,165 @@
+		};
+
+		// Store the current editor values into the state autosave.
+		dispatch( resetAutosave( toSend ) );
```
@youknowriad here when heartbeat fires, we store the current editor values into the autosave. if the values have changed since the last tick, state.autosave.isDirty will now be true, if nothing changed it will be false

adamsilverstein replied:
interestingly that wasn't the case in my testing, the dirty detection worked as expected. I may need some help here implementing this properly.

adamsilverstein replied:
@aduth can you offer a suggestion on how to refactor? I retested and it IS working as I expected it to.

I tested by logging `isPostAutosaveDirty( state ) )` before and after making some changes to the post and triggering a heartbeat:

![](https://cl.ly/1y2V1p0s011Z/Image%202018-02-20%20at%204.52.17%20PM.png)

adamsilverstein commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,165 @@
+		dispatch( resetAutosave( toSend ) );
+
+		// If the autosave is clean, no need to save.
+		if ( ! isPostAutosaveDirty( state ) ) {
```
@youknowriad this checks the isDirty status of state.autosave, if the autosave isn't dirty we don't autosave

adamsilverstein commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,166 @@
+			// Autosave is complete, success or not.
+			dispatch( toggleAutosave( false ) );
+			if ( data.wp_autosave.success ) {
+				dispatch( updateAutosaveStatusMessage( data.wp_autosave.message ) );
```
@youknowriad this action resets state.autosave.isDirty

jasmussen commented:
> I added a component named AutosaveMessage that displays the autosave message when available. I tried adding in the revisions area, but remembered that only shows when there is a stored revision. for new posts you can have autosaving but no stored revisions, so that placement wasn’t great.

> I then tried moving the status indicator right below the publish date. This looks better, but could use some design attention Here is how it looks now:

Nice work, that's probably fine for the purpose of this PR and can then be iterated later. Do we need the "at"? Can it just be "Draft saved [time]"?

adamsilverstein commented:
> Do we need the "at"? Can it just be "Draft saved [time]"?

This string is being displayed exactly as returned by WordPress core. The string is created (and passed thru translation) here: https://github.com/WordPress/WordPress/blob/aaf99e691391cfceb004d848450dbbf3344b1bee/wp-admin/includes/misc.php#L1107 so I think we would need to change it there (removing 'at' in JavaScript would break translatability)

jasmussen commented:
No need to jump through hoops. Thank you again.

adamsilverstein commented:
Fixed up tests and linting, this should be ready to merge. Can we get this in then we can keep iterating? cc: @youknowriad @aduth any feedback appreciated, this should be ready.

pento commented:
@adamsilverstein: Should this PR fix post previews not working for posts that have been published, then modified? The post preview button does an `autosave()` before opening the preview link, but it looks like it needs some more information.

adamsilverstein commented:
> Should this PR fix post previews not working for posts that have been published, then modified? 

Possibly, I'll have to test that to make sure (it may need a little adjusting) - thanks for the link to the existing issue. 

Going to review this code again today after a break from it, realized I want to make sure we really need `autosave.js`, I think we are replacing most of its functionality at this point anyway. Also, I'd like to find a better way to enqueue `heartbeat` so its not a strict dependency for Gutenberg - otherwise Gutenberg will fail to load entirely if heartbeat is disabled as raised in https://github.com/WordPress/gutenberg/issues/4547

aduth reviewed (commented):
Not to diminish the work that's been done here, but if we're trying to prove Gutenberg as a REST API based modern JavaScript app, it feels a bit odd that we're perpetuating an `admin-ajax.php` + `jQuery` autosave implementation.

adamsilverstein commented:
>  it feels a bit odd that we're perpetuating an admin-ajax.php + jQuery autosave implementation.

@aduth - this is trying to tie Gutenberg to existing core JS apis. I did explore a new/custom  implementation for autosaves in https://github.com/WordPress/gutenberg/pull/4156 - a new REST controller for autsaves was required for core because the current REST API endpoints don't directly support autosaves. Compared to that PR, this is a much simpler approach and achieves more.

There are some benefits to leveraging the existing hearbeat api autosave mechanism: autosave behavior will remain consistent with the current behavior and can be adjusted using the same mechanisms allowing you to change the heartbeat interval, disable entirely, pause/restart, localsaves,  etc...  And we'll want heartbeat active anyway to provide a persistent connection mechanism, for things like nonce refreshing, post locking as well as for back-compat as it is an existing API.

adamsilverstein commented on edit-post/components/sidebar/autosave-message/index.js:
```diff
@@ -0,0 +1,35 @@
+	}
+	return (
+		<PanelRow>
+			<span>{ __( 'Autosave' ) }</span>
```
probably not, i'll try removing

adamsilverstein replied:
removed

aduth replied:
Not removed?

adamsilverstein replied:
removed and committed this time

adamsilverstein commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,159 @@
+ * Set up the heartbeat functionality for Gutenberg.
+ */
+export function setupHeartbeat() {
+	const $document = jQuery( document );
```
no idea! its a dependency for heartbeat so I know its available, should I add a global comment to the file header?

aduth replied:
>no idea! its a dependency for heartbeat so I know its available, should I add a global comment to the file header?

First-and-foremost, we should make this error.

Then we should `import jQuery from 'jquery';`

adamsilverstein replied:
not sure how to achieve the erroring; added the import

adamsilverstein commented on editor/store/test/effects.js:
```diff
@@ -508,7 +571,7 @@ describe( 'effects', () => {
 						id: 'a9691cf9-ecaa-42bd-a9ca-49587e817647',
 						title: 'My cool block',
-						content: '<!-- wp:core/test-block {"name":"Big Bird"} /-->',
+						content: '<!-- wp:core/test-block { "name":"Big Bird" } /-->',
```
no idea, maybe i opened the file in visual studio code? will correct.

aduth replied:
>no idea, maybe i opened the file in visual studio code? will correct.

Not corrected?

jasmussen commented:
I've noticed some issues with saving, lately, and I have ZERO idea if it is related to the heartbeat API. So I recorded a very long GIF that I hope Github will accept:

![occasional autosave not firing while typing](https://user-images.githubusercontent.com/1204802/35988415-a351ff5c-0cfe-11e8-8b03-8e848205c8cd.gif)

It takes a while, probably a couple of minutes, but it illustrates the issue well. Basically, when saving has finished and you start editing an existing paragraph, this paragraph doesn't seem to be detected as changed, until you deselect the block. Which means if you close the tab before deselecting, you can lose work.

adamsilverstein commented on editor/store/selectors.js:
```diff
@@ -891,6 +914,26 @@ export function didPostSaveRequestFail( state ) {
+ * @returns {boolean}       Whether the network is connected
+ */
+export function isNetworkConnected( state ) {
+	return !! state.networkIsConnected;
```
we don't, I was following the pattern elsewhere in the file and assumed it was desired. will remove.

adamsilverstein commented on editor/store/selectors.js:
```diff
@@ -891,6 +914,26 @@ export function didPostSaveRequestFail( state ) {
+ * @returns {boolean}       Whether the post is autosaving
+ */
+export function isAutosavingPost( state ) {
+	return !! state.currentlyAutosaving;
```
I will remove the type coercion.

adamsilverstein commented on editor/store/reducer.js:
```diff
@@ -494,6 +494,64 @@ export function blocksMode( state = {}, action ) {
+ */
+export function networkIsConnected( state = true, action ) {
+	switch ( action.type ) {
+		case 'IS_CONNECTED':
```
ok, will change

adamsilverstein commented on editor/store/reducer.js:
```diff
@@ -494,6 +494,64 @@ export function blocksMode( state = {}, action ) {
+ * @param  {Object} action Dispatched action
+ * @returns {Object}        Updated state
+ */
+export const autosave = withChangeDetection( function( state = { message: '', autosave: false }, action ) {
```
null would be fine, i think we want the property defined though. I don't completely understand the comment about `withChangeDetection` I was trying to follow the docs/pattern for the use in the editor reducer - could you provide a corrected line?

adamsilverstein commented on editor/store/effects.js:
```diff
@@ -305,6 +344,11 @@ export default {
 			effects.push( resetBlocks( blocks ) );
 		}
 
+		// Resetting post should occur after blocks have been reset, since it's
```
yea, the PR is getting stale :( I'll work on a rebase or cherry pick out the duplication

adamsilverstein replied:
i think this is fixed now, not seeing duplication here

adamsilverstein commented on editor/store/effects.js:
```diff
@@ -179,6 +186,37 @@ export default {
+			) );
+		}
+	},
+	REQUEST_CONNECTION_LOST( action, store ) {
```
I'm happy to pull this out if you prefer, then I can re-add it in separate pr

adamsilverstein replied:
Please let me know if I need to separate out these changes

adamsilverstein commented on editor/store/actions.js:
```diff
@@ -36,6 +36,35 @@ export function resetPost( post ) {
+export function resetAutosave( autosaveData ) {
+	return {
+		type: 'RESET_AUTOSAVE',
+		autosave: autosaveData,
```
i did this because i was getting an 'autosave already defined' linting error, because there is an action named 'autsave' in the same scope.

adamsilverstein commented on editor/store/actions.js:
```diff
@@ -36,6 +36,35 @@ export function resetPost( post ) {
+ * Returns an action object used in signaling that the latest version of the
+ * autosave has been received.
+ *
+ * @param  {Object} autosaveData Post object
```
eagle eyes

adamsilverstein replied:
fixed along with all the linting issues in https://github.com/WordPress/gutenberg/pull/4218/commits/2ae1c94b095c4ac0ebe6206bdf2e45a59e8d8376

adamsilverstein commented on editor/components/provider/index.js:
```diff
@@ -57,6 +57,11 @@ class EditorProvider extends Component {
 		}
+
+		// Display a notice if an autosave exists.
+		if ( props.autosave ) {
```
it is not, it was lost in a merge, its supposed to be part of settings and localized at startup. Here is the original code, I will restore: https://github.com/WordPress/gutenberg/pull/4218/commits/5295bf321c285987142c2b8f8e8106ec28fc4288

adamsilverstein replied:
restored in https://github.com/WordPress/gutenberg/pull/4218/commits/2c701f7a7c16599ddc80cf5e0b81e371fc2bab7d, you can test as follows:

* create and publish a post, make some changes to the post (and blur out of the paragraph block)
* in the console do `wp.heartbeat.connectNow()`
* refresh the page, ignoring the AYS warning

you will see the autosave notice

adamsilverstein commented on edit-post/components/sidebar/autosave-message/index.js:
```diff
@@ -0,0 +1,35 @@
+	return (
+		<PanelRow>
+			<span>{ __( 'Autosave' ) }</span>
+			<div>{ message }</div>
```
sure, i thought `children` was somehow reserved or special :)

adamsilverstein replied:
renamed

adamsilverstein commented on edit-post/components/sidebar/autosave-message/index.js:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Internal dependencies
+ */
+import { getAutosaveMessage } from '../../../../editor/store/selectors';
```
i'll move to edit-post/store

adamsilverstein replied:
moved

adamsilverstein commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,159 @@
+import store from '../store';
```
right, will correct

aduth replied:
>right, will correct

Not corrected?

adamsilverstein replied:
ok, got it now i think

adamsilverstein commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,159 @@
+	 *
+	 * @since 1.9.0
+	 *
+	 * @returns {Object|boolean} postData The autosaved post data to send, or false if no autosave is needed.
```
ah, right. will rebase and clean up linting

adamsilverstein commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,159 @@
+
+		// If the autosave is clean, no need to save.
+		if ( ! isPostAutosaveDirty( state ) ) {
+			return false;
```
I'll test those out, as long as the heartbeat api accepts any falsey value I can change to null.

adamsilverstein replied:
updated

adamsilverstein commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,159 @@
+		// Block autosaving for 10 seconds.
+		wp.autosave.server.tempBlockSave();
+
+		// Dispath an event to set the state isAutosaving to true..
```
👍

aduth replied:
>👍

Still seeing typo.

adamsilverstein replied:
hmm, i think i missed a commit somewhere. fixed.

adamsilverstein commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,159 @@
+		dispatch( toggleAutosave( true ) );
+
+		// Trigger some legacy events.
+		$document.trigger( 'wpcountwords', [ toSend.content ] )
```
nothing I'm guessing the idea was to support legacy code expecting these events. Let me see how plugins in the repo use these actions, will remove for now.

aduth replied:
>will remove for now.

Not removed?

adamsilverstein replied:
i left only the legacy event firing `$document.trigger( 'before-autosave', [ toSend ] )`

aduth replied:
>i left only the legacy event firing `$document.trigger( 'before-autosave', [ toSend ] )` (plugins may hook into this)

What sorts of things are these plugins doing? Are we expecting those should still work? The [one example](https://github.com/WordPress/WordPress/blob/e6ccdf161fd1e2c9d836081d148f1fbb16767574/wp-admin/js/post.js#L212-L234) of this in existing Core code would certainly not.

adamsilverstein commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,159 @@
+		const autosaveData = save();
+
+		if ( autosaveData ) {
+			data.wp_autosave = autosaveData;
```
I don't think this contains/persists the data when its passed in - instead its an object we attach data to to include in the next heartbeat send. I'll double check. Also, I think we may want to store autosave data in state so we can reflect it in the ui (for example the last autosave time)?

aduth replied:
>Also, I think we may want to store autosave data in state so we can reflect it in the ui (for example the last autosave time)?

Are we using this or just [foresee that we need it](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)?

adamsilverstein replied:
This line is attaching the autosave data to the heartbeat request.

adamsilverstein commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,159 @@
+	/**
+	 * Disable buttons and throw a notice when the connection is lost.
+	 *
+	 * @since 1.9.0
```
i'll remove these

adamsilverstein commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,159 @@
+	$document.on( 'heartbeat-connection-lost.autosave', function( event, error, status ) {
+		// When connection is lost, keep user from submitting changes.
+		if ( 'timeout' === error || 603 === status ) {
+			dispatch( showDisconnectionNotice() );
```
Ah I see, instead of as a separate action? I'll give that a shot

aduth replied:
>Ah I see, instead of as a separate action? I'll give that a shot

How'd this go?

adamsilverstein commented:
@jasmussen this looks like a known issue (i can't find the issue # atm) that posts are not marked as dirty and thus autosavable until you blur out of a block. changes inside a block do not currently cause the post to become 'dirty' so autosaves don't happen. This behavior is not fixed by this PR as it relies on the same mechanism to detect dirty state.

jasmussen commented:
Thank you

kraftbj commented:
> current REST API endpoints don't directly support autosaves

I didn't see a trac ticket for this so I created one https://core.trac.wordpress.org/ticket/43316

adamsilverstein commented:
@aduth I have addressed all your feedback and this is ready for an additional review.

adamsilverstein commented:
@aduth @youknowriad 
Going to refresh this patch one more time because its conflicting again. I'd love to get this merged, or if you aren't going to merge it please let me know and I will withdraw it. I realize there is a ton on your plates, still - this has been open six weeks and I would appreciate a merge or rejection so I can move on to helping with other parts of gutenberg.

Thanks.

aduth commented:
Apologies for the delay in revisiting for review. I'll plan to set aside some time this weekend.

I might expect it would make for fewer conflicts and an easier refresh to squash what you have and force-push a rebased branch.

aduth commented on lib/client-assets.php:
```diff
@@ -778,6 +780,39 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
+			'_wpAutosave',
+			array(
+				'id'        => $autosave->ID,
+				'edit_link' => add_query_arg( 'gutenberg', true, get_edit_post_link( $autosave->ID ) ),
```
Should we camel-case this property name since we control its output?

adamsilverstein replied:
camelCased

aduth commented on lib/client-assets.php:
```diff
@@ -778,6 +780,39 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
+			'wp-editor',
+			'_wpAutosave',
+			array(
+				'id'        => $autosave->ID,
```
Where are we using this value?

adamsilverstein replied:
right, we can remove that.

adamsilverstein replied:
removed

aduth commented on lib/client-assets.php:
```diff
@@ -778,6 +780,39 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
+	$show_autosave = false;
+
+	// Is the autosave newer than the post?
+	if (
```
This logic is slightly more complex than I'm comfortable inlining in the script enqueue function. Could be externalized, ideally with some more comments clarifying what's being done in looping over revision fields.

adamsilverstein replied:
Ok, I can change this. Please note this code comes verbatim from core - `wp-admin/edit-form-advanced.php`

adamsilverstein replied:
extracted this

aduth commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,153 @@
+		}
+
+		// Check if its time for another autosave.
+		if ( ( new Date() ).getTime() < wp.autosave.nextRun ) {
```
Where is `nextRun` set?

adamsilverstein replied:
in wp-includes/js/autosave.js

aduth commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,153 @@
+	 */
+	const save = function() {
+		// Bail early if autosaving is suspended or saving is blocked.
+		if ( wp.autosave.isSuspended || wp.autosave._blockSave ) {
```
Where are these properties set?

adamsilverstein replied:
in autosave.js :) plugins can use the `wp.autosave` global to suspend and resume autosaving.

aduth commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,153 @@
+		// Trigger a hook action.
+		doAction( 'editor.beforeAutosave', toSend );
+
+		toSend._wpnonce = jQuery( '#_wpnonce' ).val() || '';
```
Should we be using `window.wpApiSettings.nonce` instead, if we have an interval set up to refresh that value?

adamsilverstein replied:
no, I think core is expecting a different nonce. perhaps we can add refreshing for this nonce as well.

aduth commented on editor/store/effects.js:
```diff
@@ -179,6 +186,37 @@ export default {
+		const { dispatch } = store;
+		const noticeMessage = __( 'You have lost your connection with the server, and saving has been disabled. This message will vanish once you\'ve reconnected.' );
+		dispatch( createWarningNotice(
+			<p>
```
Does a notice message need to have `<p>` or `<span>` wrappers? Ideally it shouldn't.

aduth commented on editor/store/actions.js:
```diff
@@ -275,6 +304,56 @@ export function autosave() {
+ */
+export function showAutosaveNotice( autosaveStatus ) {
+	return {
+		type: 'REQUEST_AUTOSAVE_EXISTS',
```
Action type should correspond more-or-less to action creator name. Why is `showAutosaveNotice` dispatching as `REQUEST_AUTOSAVE_EXISTS` ?

adamsilverstein replied:
renames actin type REQUEST_AUTOSAVE_NOTICE

aduth replied:
>renames actin type REQUEST_AUTOSAVE_NOTICE

Why the inconsistency between the action creator function name and the action type name?

aduth commented on editor/store/reducer.js:
```diff
@@ -582,6 +582,64 @@ export function blocksMode( state = {}, action ) {
+	}
+
+	return state;
+}, { resetTypes: [ 'UPDATE_AUTOSAVE_STATUS_MESSAGE' ] } );
```
It's not entirely obvious that `UPDATE_AUTOSAVE_STATUS_MESSAGE` is better represented as the action type which occurs as the result of a successful autosave. Maybe the status message should be updated as a separate (side-effect?) action?

adamsilverstein replied:
I'm not sure I understand what you mean "update as a separate action", how would this be implemented?

aduth commented on edit-post/components/sidebar/post-status/index.js:
```diff
@@ -35,6 +36,7 @@ function PostStatus( { isOpened, onTogglePanel } ) {
 		<PanelBody className="edit-post-post-status" title={ __( 'Status & Visibility' ) } opened={ isOpened } onToggle={ onTogglePanel }>
 			<PostVisibility />
 			<PostSchedule />
+			<AutosaveMessage />
```
Might be worth some design feedback whether we want this text here, since the rest of the panel are options configurable by the user for publishing; this is just a text status, one which may not be important for the user at all.

adamsilverstein replied:
yes I would love some design feedback and we discussed this on the PR previously... @jasmussen suggested iterating once merged although I can't find the original comment.

aduth commented on editor/utils/heartbeat.js:
```diff
@@ -0,0 +1,153 @@
+	 *
+	 * @todo update the _wpnonce used by autosaves.
+	 */
+	$document.on( 'heartbeat-tick', ( event, response ) => {
```
Should the now-duplicated logic for `heartbeat-tick` in `edit-post/index.js` be removed?

adamsilverstein replied:
yes, this was lost in a previous merge

adamsilverstein replied:
removed in https://github.com/WordPress/gutenberg/pull/4218/commits/5459e20fe5715ffab1bf0e3393045f2eefc1737c

adamsilverstein commented:
@aduth - I removed all of the code related to network connection status monitoring in https://github.com/WordPress/gutenberg/pull/4218/commits/ae8d51526ba5ee9ee1e51f212781b80fcb3eedad, I can bring that back in a future PR. I will work on addressing your feedback next.

azaozz commented:
> if we're trying to prove Gutenberg as a REST API based modern JavaScript app, it feels a bit odd that we're perpetuating an admin-ajax.php + jQuery autosave implementation.

> I didn't see a trac ticket for this so I created one https://core.trac.wordpress.org/ticket/43316

See https://core.trac.wordpress.org/ticket/43316#comment:2 and https://core.trac.wordpress.org/attachment/ticket/43316/43316.patch which is (first run of) adding autosave capability to REST API's `update_item()`.

adamsilverstein commented:
@azaozz great! having an autosave REST API endpoint is going to be a requirement for published post previews. 

We may still want to use heartbeat to transport regular autosaves since it is a well tested, established and documented API and offers useful developer hooks (and is already running anyway on the post edit screen). Or, do you think it is worth recreating the heartbeat logic in Gutenberg instead?

azaozz commented:
I tend to agree with @aduth https://github.com/WordPress/gutenberg/pull/4218#pullrequestreview-94907664. Gutenberg is a separate component, not an integral part of the Edit Post screen. It should work equally well on an (old style) page like Edit Posts, but also in single page apps. As such it's not ideal for it to depend on the old heartbeat.js, AJAX, jQuery, etc.

Heartbeat is fine-tuned to work in different conditions. It's not particularly suitable for single page apps. We may want some functionality from it, but then it will have to be rewritten to better fit the "new" JS: REST API, no dependencies on old libraries, etc. Essentially it is a polling script that can be managed from the server.

In that terms I think it will be better to make the autosave in Gutenberg use the REST API and not depend on heartbeat.js.

Also, the in-browser autosave is more important/critical than the server autosave. It ensures the user **never** looses any content (browser crash, network down, etc.). There were some ideas to not do server autosave directly, just sync the in-browser autosave data with the server. May be something to look at for v2 :)

Server autosaves are standard saves. Each time a post is saved we update the content and create a revision.  The only difference is that server autosaves are partial: either only update the content or only create or update a revision. Thinking that the best way to handle them in the REST API is to pass a flag when the save is automatic and when we specifically request an autosave.

Looking at https://github.com/WordPress/gutenberg/pull/4156, don't think we need separate controller as autosaves are standards revisions. Passing `is_autosave` when requesting a revision would be sufficient. Going to update https://core.trac.wordpress.org/ticket/43316 to add that.

adamsilverstein commented:
@azaozz Thanks for the feedback. 

I'm happy to move in that direction if the consensus is thats best - that is close to my original idea for how to handle autosaves. When I tried out using heartbeat instead I was pleased with how well it worked and thought building on an established core API made sense and added some benefits.

I'm not sure about the idea of 'passing flags' to indicate that you want to save or retrieve an autosave - that doesn't seem very 'RESTful' and makes autosaves operate opaquely. I'm curious what @rmccue thinks. 

In addition to the differences you mentioned, autosaves are also unique because one autosave is saved (at most) for each user. Also, autosaves are deleted when the post is loaded with newer content. 

Based on your feedback I am going to further reduce the scope of this ticket to only showing the warning at the top of the screen when an autosave is present. I will remove the heartbeat autosaving and we can tackle that elsewhere.

adamsilverstein commented:
@azaozz ps If we do add the code in #43315, would the plan be to backport a few versions or would Gutenberg autosaves require 5.0?

azaozz commented:
> When I tried out using heartbeat instead I was pleased with how well it worked
Yeah, it is an established, polished API. But it does polling, i.e. a lot of requests that are usually "empty". If we implement autosaves through the REST API, we will be saving only when needed.

> I'm not sure about the idea of 'passing flags' to indicate that you want to save or retrieve an autosave
The alternative is to do `post_type: autosave` and then "convert" that to `revision` before using the rest of WP. That's a bit more RESTful but is different of how autosaves are currently handled and may be somewhat confusing to developers.

> Based on your feedback I am going to further reduce the scope of this ticket to only showing the warning at the top of the screen when an autosave is present.
It will need a bit more: autosaves are triggered on an interval of 120 seconds. We need to check if the post is "dirty" for the purposes of saving to the server. That may need the same comparison as in https://core.trac.wordpress.org/browser/tags/4.9.4/src/wp-includes/js/autosave.js#L102.

It will also need a way to do immediate autosave on request and reset the interval. That's useful for previewing the very latest data from the editor and previewing changes to published posts.

adamsilverstein commented:
>It will need a bit more: autosaves are triggered on an interval of 120 seconds. We need to check if the post is "dirty" for the purposes of saving to the server. That may need the same comparison as in https://core.trac.wordpress.org/browser/tags/4.9.4/src/wp-includes/js/autosave.js#L102.

right, some of that dirty detection code is in this branch already we may be able to reuse. i mostly want to get the parts of this that work as is (autosave detection) and then start a new PR for the new approach to autosaving.

> It will also need a way to do immediate autosave on request and reset the interval. That's useful for previewing the very latest data from the editor and previewing changes to published posts.

I was thinking need to do an autosave immediately before you previewing published posts, as i recall core currently does this and then redirects to show the preview, right? So in Gutenberg: when you click preview, if the post was dirty, the autosave would fire, then when complete the preview would open (showing the autosave)

bordoni commented on lib/client-assets.php:
```diff
@@ -902,6 +937,13 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
+	$post_autosave = get_autosave_newer_than_post_save( $post );
+	if ( $post_autosave ) {
+		$editor_settings['autosave'] = array(
+				'editLink' => add_query_arg( 'gutenberg', true, get_edit_post_link( $post_autosave->ID ) ),
```
First, Dude nice work on this Pull Request.
There is an extra tab here.

adamsilverstein commented:
@bordoni thanks for the review, corrected the extra tab in https://github.com/WordPress/gutenberg/pull/4218/commits/40c8bbd63b4ca6a63aa5a88548e5bb24b77e03c0

adamsilverstein committed:
Rework PR 4218

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' into feature/heartbeat-autosaving

# Conflicts:
#	editor/components/provider/index.js

adamsilverstein committed:
switch to withDispatch approach

adamsilverstein committed:
fixes for eslint

azaozz commented on lib/client-assets.php:
```diff
@@ -799,6 +799,41 @@ function gutenberg_capture_code_editor_settings( $settings ) {
+ * @return WP_Post|boolean The post autosave. False if none found.
+ */
+function get_autosave_newer_than_post_save( $post ) {
+	// Add autosave data if it is newer and changed.
```
If we need to do that in PHP, it means the API is not working well. Why do we need to two "get" endpoints there for autosaves? They are really pointless if we can't get the latest revision from the API.

adamsilverstein replied:
Yea, we could do this with the API instead, good suggestion.

azaozz replied:
👍

azaozz commented on lib/client-assets.php:
```diff
@@ -799,6 +799,41 @@ function gutenberg_capture_code_editor_settings( $settings ) {
+	) {
+		// Iterate thru revisioned fields checking for any changes.
+		foreach ( _wp_post_revision_fields( $post ) as $autosave_field => $_autosave_field ) {
+			if ( normalize_whitespace( $autosave->$autosave_field ) != normalize_whitespace( $post->$autosave_field ) ) {
```
This is not needed any more. I know it exists in edit-form-advanced.php but it is here to catch identical autosaves when editing super old posts (from before 2.7 I think).

I'd be +1 to skip it here.

adamsilverstein replied:
do you mean the entire check for changed revisioned fields or just the normalize_whitespace part?

azaozz replied:
Uh, sorry for not being clear. Yes, the entire check. It was intended to fix existing posts that used to have identical autosave revisions at some point, very long tie ago :)

azaozz commented on lib/client-assets.php:
```diff
@@ -799,6 +799,41 @@ function gutenberg_capture_code_editor_settings( $settings ) {
+/**
+ * Retrieve a stored autosave that is newer than the post save.
+ *
+ * Deletes autosaves that are older than the post save.
```
Not true. Deleted identical autosaves only when they are newer than the post (yep, kind of pointless) :)

adamsilverstein replied:
Will adjust docblock

azaozz commented on lib/client-assets.php:
```diff
@@ -799,6 +799,41 @@ function gutenberg_capture_code_editor_settings( $settings ) {
+	}
+
+	// If this autosave isn't newer and different from the current post, remove.
+	if ( $autosave && ! $show_autosave ) {
```
This logic is different than the one in edit-form-advanced.php. Why would you want to delete older autosaves here?

adamsilverstein replied:
Trying to match this: https://core.trac.wordpress.org/browser/tags/4.9.5/src/wp-admin/edit-form-advanced.php#L217, possibly this is different.

essentially when the editor loads if the autosave isn't going to be shown (because its older) than the published post) delete it.

azaozz replied:
The whole bit of code from there:
```
if ( $autosave && mysql2date( 'U', $autosave->post_modified_gmt, false ) > mysql2date( 'U', $post->post_modified_gmt, false ) ) {
  foreach ( _wp_post_revision_fields( $post ) as $autosave_field => $_autosave_field ) {
    if ( normalize_whitespace( $autosave->$autosave_field ) != normalize_whitespace( $post->$autosave_field ) ) {
      $notice = sprintf( __( 'There is an autosave of this post that is more recent than the version below. <a href="%s">View the autosave</a>' ), get_edit_post_link( $autosave->ID ) );
      break;
    }
  }
  // If this autosave isn't different from the current post, begone.
  if ( ! $notice )
    wp_delete_post_revision( $autosave->ID );

  unset($autosave_field, $_autosave_field);
}
```

Note how the `wp_delete_post_revision( $autosave->ID );` is inside the conditional that checks if the autosave is newer than the post, and depends on all autosave fields being equal to all post fields.

As mentioned above, this is an old bit of code that is not needed any more. Can keep it if needed but the logic has to be exactly the same as in edit-form-advanced.php.

adamsilverstein replied:
Ok, thanks for clarifying. I'll bring the existing logic over exactly. I think it is worth deleting stale revisions if only for cleanup purposes.

adamsilverstein committed:
remove unneeded braces

adamsilverstein commented on editor/store/effects.js:
```diff
@@ -202,6 +205,24 @@ export default {
+				<p>
+					<span>{ noticeMessage }</span>
+					{ ' ' }
+					{ <a href={ autosaveStatus.editLink }>{ __( 'View the autosave' ) }</a> }
```
removed

adamsilverstein commented on editor/store/effects.js:
```diff
@@ -202,6 +205,24 @@ export default {
+		const noticeMessage = __( 'There is an autosave of this post that is more recent than the version below.' );
+		if ( autosaveStatus ) {
+			dispatch( createWarningNotice(
+				<p>
```
I'm following the exact patter from above in the same file:

https://github.com/WordPress/gutenberg/blob/ebe5dd4dfb64bb949dac6f3aabb13cc122497dac/editor/store/effects.js#L172-L179

Can we tackle refactoring the notices API in a separate PR?

adamsilverstein committed:
REQUEST_AUTOSAVE_NOTICE: bail early if autosaveStatus is falsey

adamsilverstein commented on editor/store/effects.js:
```diff
@@ -202,6 +205,24 @@ export default {
+	REQUEST_AUTOSAVE_NOTICE( action, store ) {
+		const { autosaveStatus } = action;
+		const { dispatch } = store;
+		const noticeMessage = __( 'There is an autosave of this post that is more recent than the version below.' );
```
Fixed

adamsilverstein commented on editor/store/effects.js:
```diff
@@ -202,6 +205,24 @@ export default {
+		const { dispatch } = store;
+		const noticeMessage = __( 'There is an autosave of this post that is more recent than the version below.' );
+		if ( autosaveStatus ) {
+			dispatch( createWarningNotice(
```
So, to clarify: returning the action here instead of dispatching it will have the same result?

adamsilverstein committed:
remove unused autosave reducer

adamsilverstein commented on editor/store/reducer.js:
```diff
@@ -901,6 +901,24 @@ export function notices( state = [], action ) {
+ */
+export function autosave( state = { message: '' }, action ) {
+	switch ( action.type ) {
+		case 'UPDATE_AUTOSAVE_STATUS_MESSAGE':
```
oops, this was a leftover; removed.

adamsilverstein commented on editor/store/effects.js:
```diff
@@ -202,6 +205,24 @@ export default {
 			__( 'Updating failed' );
 		dispatch( createErrorNotice( noticeMessage, { id: SAVE_POST_NOTICE_ID } ) );
 	},
+	REQUEST_AUTOSAVE_NOTICE( action, store ) {
```
I'm trying to follow the existing pattern here, i don't really understand what is best here so appreciate any further guidance if doing it this way isn't right.

adamsilverstein committed:
whitespace

adamsilverstein commented on lib/client-assets.php:
```diff
@@ -799,6 +799,41 @@ function gutenberg_capture_code_editor_settings( $settings ) {
 	$gutenberg_captured_code_editor_settings = $settings;
 	return false;
 }
+/**
```
fixed

adamsilverstein committed:
Merge branch 'master' into feature/heartbeat-autosaving

adamsilverstein committed:
remove field change check when retrieving autosave newer than post

adamsilverstein committed:
cleanup, doc blcoks

adamsilverstein committed:
exit early if autosave false

adamsilverstein committed:
Merge branch 'master' into feature/heartbeat-autosaving

adamsilverstein committed:
add missing simicolon

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' into feature/heartbeat-autosaving

# Conflicts:
#	editor/components/provider/index.js

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' into feature/heartbeat-autosaving

# Conflicts:
#	editor/store/effects.js

adamsilverstein committed:
spacing cleanup

adamsilverstein committed:
fix for eslint

aduth reviewed (commented):
Notices as a whole are in a bit of a messy state right now, and I'd agree we don't need to tackle all of the issues here, as they're not really specific to what you're proposing. Some problems I identify are:

- Patterns for attaching notices as effects of other behaviors (some related discussion at #6999)
- Notice handling moved outside editor to a generalized component with its own state (similar to `./viewport`)
- Consistent UX for notices with actions
   - You shouldn't need to generate the `p`, nor the `a`, this should be instead expressed as something like:

```js
createWarningNotice( noticeMessage, {
	action: {
		text: __( 'View the autosave' ),
		url: autosaveStatus.editLink,
	}
} );
```

adamsilverstein commented on editor/store/effects.js:
```diff
@@ -238,6 +241,25 @@ export default {
+	REQUEST_AUTOSAVE_NOTICE( action, store ) {
+		const { autosaveStatus } = action;
+		const { dispatch } = store;
+		if ( ! autosaveStatus ) {
```
i am defensive by default in my programming, perhaps to a fault. might be fine to omit this.

adamsilverstein commented on editor/store/effects.js:
```diff
@@ -238,6 +241,25 @@ export default {
+		const noticeMessage = __( 'There is an autosave of this post that is more recent than the version below.' );
+		dispatch( createWarningNotice(
+			<p>
+				<span>{ noticeMessage }</span>
```
good question, i can test again - my guess is this was introduced to match other existing notices.

danielbachhuber commented:
Just tested this locally. Looks like it's getting there :)

A couple of notes:

1. Autosaved published posts have `(no title)` as the autosave title:

<img width="1140" alt="image" src="https://user-images.githubusercontent.com/36432/40855222-d2158e2a-6588-11e8-93bb-a9ee8128002e.png">

I'd expect an actual title in that field.

2. For the life of me, I couldn't get an autosave created for a draft post. Is this expected behavior? It seems desirable to me, but I want to make sure it's deliberate.

More specifically, when I edit a draft post, the autosave behavior saves the post itself, and restores me to that latest version:

![autosavepost](https://user-images.githubusercontent.com/36432/40855555-b1ad8966-6589-11e8-94e0-f0d5db9c969d.gif)

danielbachhuber commented:
> Autosaved published posts have `(no title)` as the autosave title:

Ah, this is reported in #7078

adamsilverstein commented:
> For the life of me, I couldn't get an autosave created for a draft post. 

I think this is the expected behavior, let me double check and also review current behavior. For drafts, I think the difference between an autosave and clicking the 'save' button is that clicking the save button creates a revision. an autosave does not.

adamsilverstein commented:
> Autosaved published posts have (no title) as the autosave title:

This should be fixed in https://github.com/WordPress/gutenberg/pull/7092

aduth commented:
@danielbachhuber @azaozz Do you have remaining feedback here?

danielbachhuber commented:
@aduth I think we should land as-is and clean up the PHP in the next release.

youknowriad committed:
Remove useless spans

youknowriad committed:
Merge remote-tracking branch 'origin/master' into feature/heartbeat-autosaving

youknowriad committed:
Drop the showAutoSaveMessage action

youknowriad committed:
Fix unit tests after removing the span

youknowriad reviewed (approved):
LGTM 👍

youknowriad committed:
Rename autosaveStatus property just "autosave"

youknowriad commented:
Thanks all and great work @adamsilverstein
